annealing_helper_functions.py

```diff
@@ -76,3 +77,55 @@ def simulated_annealing(graph, function, initial_X, initial_temp, cool, reanneal
                 T = 1.
             
     return X, history
+
+def prob(a, b):
+    return np.exp(a / b)
+
+# Adapted from Lecture14_Parallel_Tempering_and_Emcee.ipynb
+def parallel_tempering(graph, function, initial_Xs, initial_temps, 
+                       iterr, swap_function, nswaps, nbefore):
+    # Make sure inputs are ok
+    assert(len(initial_temps) == len(initial_Xs)), "Mismatched input dimensions"
+    assert(initial_temps[0] == 1), "First temperature should be one"
+
+    # Initialize stuff
+    nsystems = len(initial_temps)
+    Xs = initial_Xs
+    Ts = initial_temps
+    P = [lambda d: np.exp(d / Ts[c]) for c in range(nsystems)]
+    prev_Es = [function(graph, Xs[i]) for i in range(nsystems)]
+    history = [[] for i in xrange(nsystems)]
+
```

Comment:
For both of them, it needs to be range rather than xrange because we actually refer to the iteration count (Ts[c], Xs[i]). For stuff where you don't refer to the iteration count, like the number of iterations in simulated_annealing(), xrange is more memory efficient. That said, using range  probably wouldn't be harmful since we're not working with huge iteration counts.